license.go

```diff
@@ -132,3 +107,112 @@ type Match struct {
 	End   int    // End offset of match in text.
 	IsURL bool   // Whether match is a URL.
 }
+
+// Type is a bit set describing the requirements imposed by a license or group of licenses.
+type Type uint
+
+const (
+	// Unrestricted indicates that the license carries no requirements.
+	// Examples: 0BSD, CC0, Unlicense.
+	Unrestricted Type = 0
+
+	// Unknown indicates that the license requirements are unknown.
+	Unknown Type = 1 << 0
+
+	// Notice indicates that the license carries notice or advertising requirements.
+	// Examples: MIT, BSD, Apache 1.1, Apache 2.0
+	Notice Type = 1 << 1
+
+	// ShareChanges indicates that source and binary distributions must include
+	// source code for any modifications made to the original code itself,
+	// made available under the same license as the original.
+	// It implies Notice.
+	// Example: LGPL.
+	ShareChanges Type = 1<<2 | Notice
+
+	// ShareProgram indicates that source and binary distributions must include
+	// source code for the entire program using the original code library,
+	// made available under the same license as the original.
+	// It implies ShareChanges.
+	// Example: GPL.
+	ShareProgram Type = 1<<3 | ShareChanges
+
+	// ShareServer indicates that any server program making use of the source code
+	// must provide a download link for the entire source code of the server.
+	// Examples: AGPL, CPAL, EUPL, SSPL.
+	// It implies ShareProgram.
+	ShareServer Type = 1<<4 | ShareProgram
+
+	// NonCommercial indicates that commercial use of the software is disallowed.
+	// Examples: CommonsClause, CC-BY-NC-4.0, Prosperity 3.0.0.
+	NonCommercial Type = 1 << 5
+
+	// Discouraged indicates that there is some legal problem with the license,
+	// making it difficult to comply with or vague about what it permits.
+	// Examples: Beerware, SISSL, WTFPL.
+	Discouraged Type = 1 << 6
+)
+
+var typeBits = []struct {
+	t Type
+	s string
+}{
+	// These implications must be listed in reverse for String to produce minimal output.
+	{ShareServer, "ShareServer"},
+	{ShareProgram, "ShareProgram"},
+	{ShareChanges, "ShareChanges"},
+	{Notice, "Notice"},
+
+	{Unrestricted, "Unrestricted"},
+	{Unknown, "Unknown"},
+	{NonCommercial, "NonCommercial"},
+	{Discouraged, "Discouraged"},
+}
+
+// String returns the type t in string form.
+// The form is a |-separated list of bit values (Unknown, Notice, and so on)
+// with a final Type(0x###) if needed.
+//
+// For example, (ShareServer|NonCommercial).String() = "ShareServer|NonCommercial".
+func (t Type) String() string {
+	if t == 0 {
+		return "Unrestricted"
+	}
+	s := ""
+	for _, b := range typeBits {
+		if b.t != 0 && t&b.t == b.t {
+			t &^= b.t
+			s += "|" + b.s
```

Comment:
doesn't this mean the string always starts with a "|"? that's not what you want, is it?

Reply:
It's easier to build that way and gets sliced away at the bottom.